Add datastore fill pattern configuration

JamesClonk opened the pull request:
Add a configuration option to define filling pattern & sort order for datastores that the vsphere-cpi picks.

JamesClonk commented:
Currently it always picks approx. the least used datastore with the most free space, but we'd like to be able to configure BOSH to choose the most used datastore, in order to fill them up efficiently.

Lafunamor commented:
This would be really helpful

nehagjain15 commented:
@JamesClonk can you please add some unit test for storage_picker and datastore_picker using this new configuration
Once you do we can run test and merge it

JamesClonk commented:
@nehagjain15 I've updated the PR accordingly. Added some unit-tests for the fill pattern configuration.
The default behaviour stays the same as before, but it now allows to configure the CPI to try and fill up the datastores more efficiently based on choosing either the most full or least full datastore to place disks.

nehagjain15 commented:
Thank you @JamesClonk.

EleanorRigby commented:
@JamesClonk : Thanks for submitting the PR. 

I have some concerns regarding the core logic of datastore picker. I have added review comments in the code. Let's discuss them first before moving on to syntactic review (if needed).

JamesClonk commented:
@EleanorRigby where can I see these comments?

nehagjain15 commented on src/vsphere_cpi/lib/cloud/vsphere/datastore_picker.rb:
```diff
@@ -44,12 +44,14 @@ def best_disk_placement(disks)
 
         if existing_ds_name =~ Regexp.new(disk.target_datastore_pattern) && datastores.keys.include?(existing_ds_name)
           datastores[existing_ds_name][:disks].push(disk)
+          break if datastore_fill_pattern != "" && datastore_fill_pattern != "none"
```
What happens with the other disks that we were looping on?

nehagjain15 commented on src/vsphere_cpi/lib/cloud/vsphere/datastore_picker.rb:
```diff
@@ -64,6 +66,8 @@ def best_disk_placement(disks)
             migration_size_accounted_for = true
           end
+
+          break if datastore_fill_pattern != "" && datastore_fill_pattern != "none"
```
I think this will violate the whole logic when few of the hosts are under maintenance mode

EleanorRigby commented:
Closing this. As the entire architecture for cluster and datastore picker has changed.
